pkg/agent/controller/serviceexternalip/controller.go

```diff
@@ -0,0 +1,451 @@
+	if err != nil {
+		return nil, fmt.Errorf("initializing service external IP assigner failed: %v", err)
+	}
+	c.ipAssigner = ipAssigner
+
+	c.serviceInformer.AddIndexers(cache.Indexers{ExternalIPIndex: func(obj interface{}) ([]string, error) {
+		service, ok := obj.(*corev1.Service)
+		if !ok {
+			return nil, fmt.Errorf("obj is not Service: %+v", obj)
+		}
+		if len(service.Status.LoadBalancer.Ingress) == 0 {
+			return nil, nil
+		}
+		return []string{service.Status.LoadBalancer.Ingress[0].IP}, nil
+	}})
+
+	c.serviceInformer.AddIndexers(cache.Indexers{externalIPPoolIndex: func(obj interface{}) ([]string, error) {
+		service, ok := obj.(*corev1.Service)
+		if !ok {
+			return nil, fmt.Errorf("obj is not Service: %+v", obj)
+		}
+		eipName, ok := service.Annotations[types.ExternalIPPoolAnnotationKey]
+		if !ok {
+			return nil, nil
+		}
+		return []string{eipName}, nil
+	}})
+
+	c.serviceInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: c.enqueueService,
+			UpdateFunc: func(old, cur interface{}) {
+				c.enqueueService(cur)
+			},
+			DeleteFunc: c.enqueueService,
+		},
+		resyncPeriod,
+	)
+
+	c.endpointsInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: c.enqueueServiceForEndpoints,
+			UpdateFunc: func(old, cur interface{}) {
+				c.enqueueServiceForEndpoints(cur)
+			},
+			DeleteFunc: c.enqueueServiceForEndpoints,
+		},
+		resyncPeriod,
+	)
+
+	c.localIPDetector.AddEventHandler(c.onLocalIPUpdate)
+	c.cluster.AddClusterEventHandler(c.enqueueServicesByExternalIPPool)
+	return c, nil
+}
+
+func (c *ServiceExternalIPController) enqueueService(obj interface{}) {
+	service, ok := obj.(*corev1.Service)
+	if !ok {
+		deletedState, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			klog.Errorf("Received unexpected object: %v", obj)
+			return
+		}
+		service, ok = deletedState.Obj.(*corev1.Service)
+		if !ok {
+			klog.Errorf("DeletedFinalStateUnknown contains non-Service object: %v", deletedState.Obj)
+			return
+		}
+	}
+	c.queue.Add(apimachinerytypes.NamespacedName{
+		Namespace: service.Namespace,
+		Name:      service.Name,
+	})
+}
+
+func (c *ServiceExternalIPController) enqueueServiceForEndpoints(obj interface{}) {
+	endpoints, ok := obj.(*corev1.Endpoints)
+	if !ok {
+		deletedState, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			klog.Errorf("Received unexpected object: %v", obj)
+			return
+		}
+		endpoints, ok = deletedState.Obj.(*corev1.Endpoints)
+		if !ok {
+			klog.Errorf("DeletedFinalStateUnknown contains non-Endpoint object: %v", deletedState.Obj)
+			return
+		}
+	}
+	service, err := c.serviceLister.Services(endpoints.Namespace).Get(endpoints.Name)
+	if err != nil {
+		klog.ErrorS(err, "failed to get Service for Endpoint", "namespace", endpoints.Namespace, "name", endpoints.Name)
+		return
+	}
+	// we only care services with ServiceExternalTrafficPolicy setting to local
```

Comment:
I don't quite understand, could you explain this a little more?
The policy can be changed from local to cluster and from cluster to local, right?